app/services/db/monthly_sp_auth_count/unique_monthly_auth_counts_by_partner.rb

```diff
@@ -0,0 +1,144 @@
+# frozen_string_literal: true
+
+module Db
+  module MonthlySpAuthCount
+    module UniqueMonthlyAuthCountsByPartner
+      extend Reports::QueryHelpers
+
+      module_function
+
+      # @param [String] key label for billing (Partner requesting agency)
+      # @param [Array<String>] issuers issuers for the iaa
+      # @param [Date] start_date iaa start date
+      # @param [Date] end_date iaa end date
+      # @return [PG::Result, Array]
+      def call(key:, issuers:, start_date:, end_date:)
+        date_range = start_date...end_date
+
+        return [] if !date_range || issuers.blank?
+
+        # Query a month at a time, to keep query time/result size fairly reasonable
+        # The results are rows with [user_id, ial, auth_count, year_month]
+        months = Reports::MonthHelper.months(date_range)
+        queries = build_queries(issuers: issuers, months: months)
+
+        ial_to_year_month_to_users = Hash.new do |ial_h, ial_k|
+          ial_h[ial_k] = Hash.new { |ym_h, ym_k| ym_h[ym_k] = Multiset.new }
+        end
+
+        queries.each do |query|
+          temp_copy = ial_to_year_month_to_users.deep_dup
+
+          with_retries(
+            max_tries: 3,
+            rescue: [
+              ActiveRecord::SerializationFailure,
+              PG::ConnectionBad,
+              PG::TRSerializationFailure,
+              PG::UnableToSend,
+            ],
+            handler: proc do
+              ial_to_year_month_to_users = temp_copy
+              ActiveRecord::Base.connection.reconnect!
+            end,
+          ) do
+            Reports::BaseReport.transaction_with_timeout do
+              ActiveRecord::Base.connection.execute(query).each do |row|
+                user_id = row['user_id']
+                year_month = row['year_month']
+                ial = row['ial']
+                if row['min_returned_at'].nil? 
+                  next
+                elsif row['max_profile_verified_at'].nil?
+                  profile_age = -1
```

Comment:
what if we used `nil` in place of `-1`?

Reply:
We were trying to account for a potential bug of min_returned_at being greater than the max_profile_verified_at date, it would "catch" both failures. We are unsure how the multiset would handle nils, and if using nil we would have to put in some more code to catch the case we mentioned

Reply:
before, the multiset was *just* user IDs right? I think it might be better to keep it that way, and have a separate hash or something to look up the profile age per user, because for the multiset to work, each month's timestamps would have to be identitcal to correctly match

Reply:
That is right, before our edits the multiset just contained user_id values.
The multiset just accounts for auth counts which is already being generated from the counts_by_iaa right? 
Though we were thinking about just getting rid of the multiset altogether for this but weren't sure if it would break other things. The userid is inside the multiset, and profile age is tied to the user so we figured it had to be tied to whatever object the userid was in.

Reply:
@zachmargolis, the mulitset was not _just_ user IDs before - it was "user IDs and auth_count".  Now we are doing "user IDs and profile_age" instead.

Reply:
> @zachmargolis, the mulitset was not _just_ user IDs before - it was "user IDs and auth_count".

To clarify: the content of the multiset before was just user IDs. And because multisets have counts per item, it can produce auth counts. But the contents of the multiset were just plain strings before

Reply:
Resolving this as we have moved away from multisets?